### PR TITLE
[runtime-security] Increase dentry resolver limits and introduce eRPC based resolution

### PR DESCRIPTION
**Pull request by Gui774ume**
### What does this PR do?

This PR introduces 2 changes:

- the dentry resolver is no longer limited to segments of 128 characters and paths of depth 16. Each segment of a path can now be up to 255 characters (which is the kernel limit) and their maximum depth is 1500.
- 2 new eRPC operations were added: `RESOLVE_SEGMENT_OP` can be used to resolve a segment and `RESOLVE_PATH_OP` can be used to resolve a full path. Both method rely on the `bpf_probe_write_user` helper to write in a user space buffer the response. A pid filter ensures that only system-probe is able to make those requests (thus our eBPF programs can write only in the user space memory of system-probe).

### Motivation

The limits of the previous resolver were reached repeatedly in staging / prod, thus yielding incorrect paths in our security alerts. Moreover the previous user space resolution required an huge amount of bpf syscalls which was suboptimal.

### Describe your test plan

Functional tests were added to ensure that both resolution methods are still working and that the old limits no longer apply.

### CODE REVIEW

File: pkg/security/ebpf/c/erpc.h [hunk 1]
```diff
@@ -55,15 +57,16 @@ int __attribute__((always_inline)) handle_discard_pid(void *data) {
...
 
-    LOAD_CONSTANT("erpc_fd", fd);
+//    LOAD_CONSTANT("erpc_fd", fd);
     LOAD_CONSTANT("erpc_pid", pid);
 
-    u32 vfs_fd = PT_REGS_PARM2(ctx);
-    if (!vfs_fd || (u64)vfs_fd != fd) {
-        return 0;
-    }
+//    u32 vfs_fd = PT_REGS_PARM2(ctx);
```
Comment (lebauce) [inline pkg/security/ebpf/c/erpc.h:66]: Why removing this check ?

File: pkg/security/probe/dentry_resolver.go [hunk 2]
```diff
@@ -268,10 +277,129 @@ func (dr *DentryResolver) ResolveFromMap(mountID uint32, inode uint64, pathID ui
...
+
+	if err = dr.erpc.Request(&dr.erpcRequest); err != nil {
+		return "", err
+	}
+
+	var keys []PathKey
+	var segments []string
+
+	i := 0
+	for i < dr.erpcSegmentSize-17 {
```
Comment (lebauce) [inline pkg/security/probe/dentry_resolver.go:341]: Is 17 sizeof(pathID) + 1 ?

File: pkg/security/ebpf/c/erpc.h (hunk 1 above)
Comment (Gui774ume) [reply pkg/security/ebpf/c/erpc.h:66]: The eRPC clients were initialized at 2 different places in the code. Updated and fixed ;)

File: pkg/security/probe/dentry_resolver.go (hunk 2 above)
Comment (Gui774ume) [reply pkg/security/probe/dentry_resolver.go:341]: Yes, I added a comment :)

Comment (Gui774ume) [reply pkg/security/probe/dentry_resolver.go:341]: Also I think it's supposed to be `<=` unless I forgot an edge case 🤔

File: pkg/config/config.go [hunk 3]
```diff
@@ -823,6 +823,8 @@ func InitConfig(config Config) {
 	// Datadog security agent (runtime)
 	config.BindEnvAndSetDefault("runtime_security_config.enabled", false)
 	config.SetKnown("runtime_security_config.fim_enabled")
+	config.BindEnvAndSetDefault("runtime_security_config.erpc_dentry_resolution_enabled", true)
```
Comment (safchain) [inline pkg/config/config.go:826]: What if one disable both ?

File: pkg/security/ebpf/c/dentry_resolver.h [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+    .key_size = sizeof(struct path_key_t),
+    .value_size = sizeof(struct path_leaf_t),
+    .max_entries = 64000,
+    .pinning = 0,
+    .namespace = "",
+};
+
+#define DR_NO_CALLBACK                              -1
+
+#define DR_OPEN_CALLBACK_KPROBE_KEY                 1
```
Comment (safchain) [inline pkg/security/ebpf/c/dentry_resolver.h:39]: enum ?

File: pkg/security/ebpf/c/dentry_resolver.h [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+#define DR_RENAME_CALLBACK_KPROBE_KEY               10
+
+struct bpf_map_def SEC("maps/dentry_resolver_kprobe_callbacks") dentry_resolver_kprobe_callbacks = {
+    .type = BPF_MAP_TYPE_PROG_ARRAY,
+    .key_size = sizeof(u32),
+    .value_size = sizeof(u32),
+    .max_entries = EVENT_MAX,
+};
+
+#define DR_OPEN_CALLBACK_TRACEPOINT_KEY       1
```
Comment (safchain) [inline pkg/security/ebpf/c/dentry_resolver.h:57]: ditto

File: pkg/security/ebpf/c/dentry_resolver.h [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+
+    if (syscall->resolver.callback >= 0) {
+        bpf_tail_call(ctx, &dentry_resolver_kprobe_callbacks, syscall->resolver.callback);
+    }
+    return 0;
+}
+
+SEC("tracepoint/dentry_resolver_kern")
+int tracepoint__dentry_resolver_kern(void *ctx) {
+    struct syscall_cache_t *syscall = peek_syscall(EVENT_ANY);
```
Comment (safchain) [inline pkg/security/ebpf/c/dentry_resolver.h:189]: it seems to be the exact same code as the kprobe one, could we make a common function called by both ?

File: pkg/security/ebpf/c/dentry_resolver.h [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+
+#pragma unroll
+    for (int i = 0; i < DR_MAX_ITERATION_DEPTH; i++)
+    {
+        map_value = bpf_map_lookup_elem(&pathnames, &state->key);
+        if (map_value == NULL)
+            goto exit;
+
+        // make sure we do not write outside of the provided buffer
+        if (state->cursor + 16 >= state->buffer_size)
```
Comment (safchain) [inline pkg/security/ebpf/c/dentry_resolver.h:248]: s/16/sizeof(state->key)/ ? does it make sense ?

File: pkg/security/module/module.go [hunk 8]
```diff
@@ -359,7 +359,6 @@ func NewModule(cfg *config.Config) (module.Module, error) {
 
 	// custom limiters
 	limits := make(map[rules.RuleID]Limit)
-	limits[sprobe.AbnormalPathRuleID] = Limit{Limit: 0, Burst: 0}
```
Comment (safchain) [inline pkg/security/module/module.go:362]: I would keep this line until we are sure to know 1. the number of event 2. what to do with them ?

File: pkg/config/config.go (hunk 3 above)
Comment (Gui774ume) [reply pkg/config/config.go:848]: Added some logic to make sure that the map resolution is set to `true` if both options are set to `false` 👍🏻

File: pkg/security/ebpf/c/dentry_resolver.h (hunk 6 above)
Comment (Gui774ume) [reply pkg/security/ebpf/c/dentry_resolver.h:189]: It will have to be a macro unfortunately: from the few tests I've made you can't have a `bpf_tail_call` on a program map that contains programs from a different type than the calling programs (even behind a if that will never be called, I guess the dead code elimination part of the verifier is done after the map coherence check ? :idk: ).

TL;DR that's why I duplicated the code, but a macro should do the trick ! Working on it ...

File: pkg/security/ebpf/c/dentry_resolver.h (hunk 4 above)
Comment (Gui774ume) [reply pkg/security/ebpf/c/dentry_resolver.h:39]: 👍🏻

File: pkg/security/ebpf/c/dentry_resolver.h (hunk 5 above)
Comment (Gui774ume) [reply pkg/security/ebpf/c/dentry_resolver.h:57]: 👍🏻

File: pkg/security/ebpf/c/dentry_resolver.h (hunk 7 above)
Comment (Gui774ume) [reply pkg/security/ebpf/c/dentry_resolver.h:248]: 👍🏻

File: pkg/security/ebpf/c/dentry_resolver.h (hunk 6 above)
Comment (Gui774ume) [reply pkg/security/ebpf/c/dentry_resolver.h:189]: 👍🏻

File: pkg/security/module/module.go (hunk 8 above)
Comment (Gui774ume) [reply pkg/security/module/module.go:362]: I think I fixed the bug I was talking about last week. Making a custom deployment in a bit to make sure. If I don't see any abnormal_path, then I think we should remove the rate limiter.

**Review by safchain — APPROVED** [review]

**Review by lebauce — APPROVED** [review]
